Updating OrientDB driver to the latest released version 3.1.x

dritter-sap opened the pull request:
Updating OrientDB driver to the latest released version 3.1.x:

- Switches to ODB version `3.1.3`
- Removes old ODB version support

Chrisdrinkingcaffee commented:
And when I try to run "./bin/ycsb load orientdb -s -P workloads/workloada -p recordcount=10000 -p table=11_a_usertable_14 -p orientdb.url=plocal:/usr/data1/orientdb-3.1.2/databases/ycsb"
It shows Storage 'plocal:/usr/data1/orientdb-3.1.2/databases/**ycsb/ycsb**' is created under OrientDB distribution' rather than "plocal:/usr/data1/orientdb-3.1.2/databases/ycsb’

dritter-sap commented:
Thanks, let me check that myself. Did you try with `orientdb.url=remote:localhost/ycsb` or just `./bin/ycsb load orientdb -s -P workloads/workloada`. Does this work?

aka-chris commented:
No, it doesn't. When I run "./bin/ycsb load orientdb -s -P workloads/workloada". It shows" Suppressed: com.orientechnologies.orient.core.exception.OSecurityAccessException: Wrong user/password to [connect] to the remote OrientDB Server instance".  And When I try "-p orientdb.user=root. It shows the same. I am sure How about you?

dritter-sap commented:
There was indeed a small issue with the transactions during insert. This should be fixed now. I tested the following successfully: starting OrientDB `docker run -d --name orientdb -p 2424:2424 -p 2480:2480 -e ORIENTDB_ROOT_PASSWORD=admin orientdb:latest` for workbooks:

workloada:
```
./bin/ycsb load orientdb -s -P workloads/workloada -p recordcount=10000 -p table=11_a_usertable_14 -p orientdb.url=remote:localhost/ycsb -p orientdb.user=admin -p orientdb.password=admin
```
workloadb:
```
./bin/ycsb load orientdb -s -P workloads/workloadb -p recordcount=10000 -p table=11_a_usertable_14 -p orientdb.url=remote:localhost/ycsb -p orientdb.user=admin -p orientdb.password=admin
```
workloadc:
```
./bin/ycsb load orientdb -s -P workloads/workloadc -p recordcount=10000 -p table=11_a_usertable_14 -p orientdb.url=remote:localhost/ycsb -p orientdb.user=admin -p orientdb.password=admin
```
workloadd:
```
./bin/ycsb load orientdb -s -P workloads/workloadd -p recordcount=10000 -p table=11_a_usertable_14 -p orientdb.url=remote:localhost/ycsb -p orientdb.user=admin -p orientdb.password=admin
```
workloade:
```
./bin/ycsb load orientdb -s -P workloads/workloade -p recordcount=10000 -p table=11_a_usertable_14 -p orientdb.url=remote:localhost/ycsb -p orientdb.user=admin -p orientdb.password=admin
```
workloadf:
```
./bin/ycsb load orientdb -s -P workloads/workloadf -p recordcount=10000 -p table=11_a_usertable_14 -p orientdb.url=remote:localhost/ycsb -p orientdb.user=admin -p orientdb.password=admin
```

Storage and tables are created correctly. Could you kindly double-check. Make sure that you use `orientdb.url=remote:localhost/ycsb` as mentioned above.

dritter-sap commented:
Hi @aka-chris, did that work for you as well, now?

aka-chris commented:
Sorry for the late reply to you. I just run this "./bin/ycsb load orientdb -s -P workloads/workloada -p recordcount=10000 -p table=11_a_usertable_14 -p orientdb.url=remote:localhost/ycsb -p orientdb.user=admin -p orientdb.password=admin" But it creates two table , which are usertable and 11_a_usertable_14. 
![image](https://user-images.githubusercontent.com/27014963/97308033-5cbfc000-189b-11eb-9c32-c7efe394a96b.png)

aka-chris commented:
![image](https://user-images.githubusercontent.com/27014963/97310208-ec666e00-189d-11eb-9743-99b61ae134ab.png)
![image](https://user-images.githubusercontent.com/27014963/97310272-fbe5b700-189d-11eb-9910-2b63cb68ed39.png)
And the indexes of the table "11_a_usertable_14" is not created.

dritter-sap commented:
It seems that the `table` parameter was also not supported in the old version of the ODB driver. So currently `usertable` is the only one table that is created and it has a hash index. If `table` is an official parameter to be supported, then I can add it.

aka-chris commented:
OK! Thank you.

dritter-sap commented:
No worries, could you merge the PR, if everything is alright then?

aka-chris commented:
Sorry, I(embarrassed) am new here but how to approve and merge the PR.

aka-chris reviewed (commented):
Nice!

dritter-sap commented:
Thanks, after merging, you can also close issue https://github.com/brianfrankcooper/YCSB/issues/1470

aka-chris commented:
sorry but how to approve and merge the PR

dritter-sap commented:
It looks like you already approved the PR. Now on the bottom of this page, you should have the possibility to `merge`. You will be asked to confirm when you click on the merge button.

aka-chris commented:
There is no this button `merge` on the bottom of this page. I swear.

aka-chris commented:
![image](https://user-images.githubusercontent.com/27014963/97323504-d19af600-18ab-11eb-916c-36372eb0586a.png)

dritter-sap commented:
Then you are not a maintainer of this repository and we will have to check with one of them. I naturally assumed that you were, since "constructive feedback" is part of the service here :) (cf. https://github.com/brianfrankcooper/YCSB/blob/master/CONTRIBUTING.md)
Let me drop an e-mail to ycsb-users@yahoogroups.com then. Maybe some maintainers listen on that channel.

aka-chris commented:
Haha. I'm just new here.  Thank you so much for your abilities and patience.

busbey commented on orientdb/configs/config.sh:
```diff
@@ -0,0 +1,29 @@
+if [ -z "$workloads" ]; then
```
need a license header.

dritter-sap replied:
done

busbey commented on orientdb/plots/collect_data.sh:
```diff
@@ -0,0 +1,116 @@
+#!/bin/bash
```
need a license header.

dritter-sap replied:
done

busbey commented on orientdb/plots/merge_results.sh:
```diff
@@ -0,0 +1,47 @@
+#!/bin/bash
```
need a license header.

dritter-sap replied:
done

busbey commented on orientdb/plots/plot.sh:
```diff
@@ -0,0 +1,139 @@
+#!/bin/bash
```
need a license header.

dritter-sap replied:
done

busbey commented on orientdb/plots/tool.sh:
```diff
@@ -0,0 +1,38 @@
+#!/bin/bash
```
need a license header.

dritter-sap replied:
done

busbey commented on orientdb/plots/update_merged.sh:
```diff
@@ -0,0 +1,32 @@
+#!/bin/bash
```
need a license header.

dritter-sap replied:
done

busbey commented on orientdb/src/main/java/site/ycsb/db/OrientDBClient.java:
```diff
@@ -1,272 +1,364 @@
 /**
  * Copyright (c) 2012 - 2016 YCSB contributors. All rights reserved.
```
update to add 2020 to the copyrights

dritter-sap replied:
done - and I meanwhile updated all to 2021, too. Kindly merge, before we get into 2022.

busbey commented on orientdb/src/main/java/site/ycsb/db/OrientDBClient.java:
```diff
@@ -1,272 +1,364 @@
+      poolCfg.addConfig(OGlobalConfiguration.DB_POOL_MAX, 10);
+      final OrientDBConfig oriendDBconfig = poolCfg.build();
+      if (orient == null) {
+        orient = new OrientDB(url, "root", "admin", oriendDBconfig);
```
This user/password should come from the ConnectionProperties we made before.

dritter-sap replied:
done

busbey commented on orientdb/src/main/java/site/ycsb/db/OrientDBClient.java:
```diff
@@ -276,48 +368,59 @@ public Status update(String table, String key, Map<String, ByteIterator> values)
-      LOG.warn("OrientDB scan operation is not implemented for remote database connections.");
-      return Status.NOT_IMPLEMENTED;
-    }
+  private String escapeQuotes(final String key) {
+    // Brutal
+    return key.replace("\"", "").replace("'", "").replace("\\", "");
+  }
```
add a javadoc to this method on what it's doing?

dritter-sap replied:
done - now called `escapeUnsupportedChars`

busbey commented on orientdb/src/test/java/site/ycsb/db/OrientDBClientBatchTest.java:
```diff
@@ -0,0 +1,105 @@
+package site.ycsb.db;
```
needs a license header

dritter-sap replied:
done

busbey commented on orientdb/README.md:
````diff
@@ -64,7 +69,29 @@ WARNING: Creating a new database will be done safely with multiple threads on a
 
-## Known Issues
 
+## Plot results
+
+To plot workload results install gnuplot, then create a directory to collect the outputs
+```
+mkdir -p orientdb/outputs/orientdb
+```
+load the data
+```
+./bin/ycsb load orientdb -s -P workloads/workloadc 
+```
+and run the workload, e.g. `workloadc`, by following the naming convention for the file name `outputRun_<workload>.txt`:
+```
+./bin/ycsb run orientdb -s -P workloads/workloadc > orientdb/outputs/orientdb/outputRun_workloadc.txt
+```
+
+After a successful worload run, goto the `plots` directory `cd orientdb/plots` and run
+```
+./tool.sh "plot" "save" ../outputs/orientdb/
+```
+The diagrams will be in the `images` directory.
+
````
this is a nice addition, but it will be confusing to folks who download the orientdb specific tarball because these directories won't be present.

you can check the tarball like this:

```
10335-busbey:YCSB busbey$ mvn -pl orientdb -am verify
...
10335-busbey:YCSB busbey$ tar tzf orientdb/target/ycsb-orientdb-binding-0.18.0-SNAPSHOT.tar.gz 
...
```

What if we pulled this tooling out of the orientdb module and generalized it for ycsb runs?

dritter-sap replied:
sounds good, but it should stay on ODB level in this PR. We could put a more general support into a new PR. Wdyt?

busbey commented on orientdb/README.md:
````diff
@@ -64,7 +69,29 @@ WARNING: Creating a new database will be done safely with multiple threads on a
+```
+mkdir -p orientdb/outputs/orientdb
+```
+load the data
+```
+./bin/ycsb load orientdb -s -P workloads/workloadc 
+```
+and run the workload, e.g. `workloadc`, by following the naming convention for the file name `outputRun_<workload>.txt`:
+```
+./bin/ycsb run orientdb -s -P workloads/workloadc > orientdb/outputs/orientdb/outputRun_workloadc.txt
+```
````
this looks like it's parsing out the default text measurements. it would be simpler to use the `exporter` settings to write e.g. a json file via `JSONArrayMeasurementsExporter` or `JSONMeasurementsExporter`

dritter-sap commented:
Hi @busbey, thanks for your review. I added the license headers, changed `escapeQuotes` to `escapeUnsupportedChars` and added docs, and made additional server user/password configurable.

I did not want to make a big deal out of the plotting, since I just needed it and wanted to reuse it next time. However, I like the idea of using the plotting support for the whole benchmark and have nothing against it to be moved to the top level.
Maybe we could move and adapt it after this PR is done?

I have some further ideas on how to slightly improve the benchmark as well as questions. Probably also something for later. Do you have a Git Discusisons board or any other way to discuss new ideas / questions?

dritter-sap commented:
Hi @busbey, are the changes sufficient - do you have further changes?

dritter-sap commented:
Hi @busbey, I responded to your comments and updated the license to 2021. Could you have a look and merge the PR?

dritter-sap commented:
@busbey @brianfrankcooper: could you kindly merge the PR. The review was answered to a year ago.
